### DISCUSSION TITLE
1BRC SIMD, targeting both Java and C via Singeli

### DISCUSSION THREAD

**Discussion by dzaima** [gunnarmorling/1brc#267] [Show and tell]
https://github.com/dzaima/1brc

Core SIMD logic written in [Singeli](https://github.com/mlochbaum/Singeli), from which either Java `jdk.incubator.vector` usage or C AVX2 intrinsics are generated (repo comes with the code of both pregenerated so Singeli isn't required to run it).

The core hot code is pretty much equivalent between C and Java, so this effectively shows the difference between the codegen of OpenJDK 21 and C (I'm testing on `clang-17`).

As I only have 8GB RAM, I'm testing on a 100M-record input (i.e. 1.38GB) on my i3-4160 (Haswell, 3.6GHz), configured to run on 2 threads. C takes 0.56s, and Java - 3.3s (14.9s if ran 10 times in a loop in the same VM, simulating 1e9 records).

Some differences between the two:

- C version uses a `vpgatherqq` to gather four 8-byte segments of the input for temperatures, but, while Java's `Vector` does have gathers, they're only for aligned loads in same-type arrays, which don't work here (we need to load arbitrary-alignment `long`s within a `byte[]`), so it's emulated with four vector loads, stores to a temp buffer, and another load to construct the vector.

- Smaller things include using a `max(a-b,0)` instead of an unsigned saturating subtract, more manual digit multiplying where C uses a single `vpmuldq`, and a couple more similar things.

But most of the perf difference should just be that OpenJDK generates less optimal machine code and does bounds checks.

**Comment by AlexanderYastrebov** [comment]
Great stuff!

Some test cases do not pass though
```
$ for sample in $(ls $BRCHOME/src/test/resources/samples/*.txt); do echo $(basename $sample); diff <(make run file=$sample | $BRCHOME/t
ocsv.sh) <($BRCHOME/tocsv.sh < ${sample%.txt}.out) ; done
measurements-10000-unique-keys.txt
measurements-10.txt
measurements-1.txt
measurements-20.txt
5c5
< Berlin1️⃣🐝🏎️;-0.3;-0.3;0.0
---
> Berlin1️⃣🐝🏎️;-0.3;-0.3;-0.3
15c15
< Nashville1️⃣🐝🏎️;-4.9;-4.9;0.0
---
> Nashville1️⃣🐝🏎️;-4.9;-4.9;-4.9
measurements-2.txt
measurements-3.txt
measurements-boundaries.txt
1c1
< Bosaso;-99.9;-99.9;0.0
---
> Bosaso;-99.9;-99.9;-99.9
measurements-complex-utf8.txt
```

**Reply by dzaima** [reply]
Interesting, will look into that. Didn't test much myself.

**Reply by dzaima** [reply]
Pushed a fix - missed initialization of the starting max temperature. (and also fixed a C version bug of an `some_uint64-1 < 0` not replaced after being ported from Java)

**Comment by AlexanderYastrebov** [comment]
Could you try it on 1e9 records file?
On my machine (https://github.com/gunnarmorling/1brc/discussions/67) it completes in 10m winning the slowest contest :rofl: 
```
real    10m10,581s
user    72m29,993s
sys     0m24,896s
```

**Reply by dzaima** [reply]
Huh. Ran it on 1e9 lines, even with HDD speeds it took `real 5m39.617s`, `user 1m44.004s`.
For that slow of a result I'd guess that the vector operations failed to JIT to actual SIMD instructions, or something to that effect; your system should have AVX2 though, so I have no clue why that'd be (I'm on `openjdk 21.0.1 2023-10-17`). Can't really say much though, this is my first time using Java's `Vector` stuff, or Java microoptimization in general.

**Reply by AlexanderYastrebov** [reply]
Thank you, the idea is great anyways, maybe something is off on my machine.
`time ./a.out` completes in 40s
```
real    0m40,298s
user    2m4,463s
sys     0m9,363s
```

**Comment by lehuyduc** [comment]
I ran your code on a 2950X, running at 8 threads (default in your code). Same machine as here https://github.com/gunnarmorling/1brc/discussions/138 , and you can download the file from my repo (13.8 GB, 1B lines)
```
real	0m20.653s
user	1m24.287s
sys	0m1.592s
```

**Reply by dzaima** [reply]
On my system currently my solution is slightly faster than yours :) (the native (`a.out`) one that is, not Java; Java's still behind by ~2.4x; I wouldn't be surprised if my solution was slower on your machine though, Zen+ has low throughput for some ops I use a lot)

**Reply by lehuyduc** [reply]
Could you post the time please? My code would have to be at least 20x slower on your CPU, which is weird (on 2950X at 8 threads it runs in ~1.5 sec)

**Reply by dzaima** [reply]
I'm testing both my and your solutions on a 1.38GB file, 2 threads, yours being 0.73s real, 1.32s user (compared to my 0.61s real, 1.22 user); `0.73/2*8 * 10 = 1.8s` is what it would be like on my CPU speeds but 8 threads and the full 10x larger file, which is aproximately the 1.5s you're seeing.

**Reply by lehuyduc** [reply]
I ran your code again on AMD 75F3 (Zen 3) because you say it's due to Zen+. The result with 8 threads (run multiple time):
Latest commit ([29879c6](https://github.com/dzaima/1brc/commit/29879c632517ce812fa80f0bc0957eecdf6e1f46) )
```
real    0m10.671s
user    0m43.381s
sys     0m3.338s

real    0m10.653s
user    0m43.289s
sys     0m3.163s
```

Commit I used to test the first time (`47216f9` fix lock issues)
```
real    0m23.421s
user    1m39.149s
sys     0m3.156s

real    0m23.504s
user    1m39.371s
sys     0m3.215s
```

My `1brc_valid16` 
```
75F3
real    0m1.893s
user    0m11.465s
sys     0m1.020s

2950X
real    0m1.434s
user    0m9.677s
sys     0m0.540s
```

So it's slow not because of Zen+.

**Reply by dzaima** [reply]
Hmm. Is that really the native version? (Java one is invoked with `make run file=...`, native is `make a.out` and then `./a.out the-measurements.txt`). Otherwise this is quite odd (The Zen+ comment was about a potential like ~10-20% difference, not 20x).

**Reply by lehuyduc** [reply]
Yes it's the native version. 
```
make a.out
./a.out measurements.txt
```
Did 100M lines of my `measurements.txt` finish in 0.6s on your computer with 2 threads? On did you use another test file?
It's very interesting result if it's the same test file

**Reply by dzaima** [reply]
The 100M line test is one created with `./create_measurements.sh 100000000` some time ago, not your input, but yes, two threads, `THREADS_1BRC=2 ./a.out measurements_1.38GB.txt`. I'll try first 100M lines of your input once it finishes downloading

**Reply by dzaima** [reply]
```bash
$ head -n100000000 measurements.txt > measurements_lehuyduc_100M.txt
$ wc measurements_lehuyduc_100M.txt
 100000000  117920945 1379537380 measurements_lehuyduc_100M.txt
$ sha256sum measurements_lehuyduc_100M.txt
981ac209f288775dbfdf0c6805cfaf53da3026aa12cfe5c34a0d6a6ba6bc66d7  measurements_lehuyduc_100M.txt
$ make a.out; time THREADS_1BRC=2 ./a.out measurements_lehuyduc_100M.txt q
make: 'a.out' is up to date.

real  0m0.596s
user  0m1.080s
sys 0m0.100s
```
(random unrelated note - in your `munmap` you probably want `file_size + 32` to match the `mmap`)

**Reply by dzaima** [reply]
Perhaps it's a kernel or similar difference (I'm on kernel 6.5.0, linux mint)? I use `fork()` with `MAP_SHARED` mappings for multithreading, whereas yours uses threads. Adding an explicit `munmap` to my solution doesn't change any timings (additionally, on my 2c4t system on the 1.38GB input a `THREADS_1BRC=256` test takes only ~80ms longer than a `THREADS_1BRC=4` one) but maybe on your systems something behaves differently. Would be quite weird though. But otherwise, our solutions should clearly have pretty comparable timings, both doing SIMD semicolon searching, SIMD name masking & hashing, and some basic hashmap stuff (only other big difference is my solution doing vectorized number parsing & using `vpgatherqq` within that, but that really shouldn't be an issue)
edit: tested on kernel 5.15.0 too, same timings. ¯\\\_(ツ)\_/¯

**Reply by lehuyduc** [reply]
Your latest code on github:

```
real	0m8.214s
user	0m35.032s
sys	0m1.554s
```

`perf stat -d ./a.out measurements.txt` 
```
 Performance counter stats for './a.out':

         35,864.80 msec task-clock                #    4.429 CPUs utilized          
             3,398      context-switches          #    0.095 K/sec                  
                10      cpu-migrations            #    0.000 K/sec                  
           854,185      page-faults               #    0.024 M/sec                  
   148,538,253,922      cycles                    #    4.142 GHz                      (37.43%)
     1,271,348,810      stalled-cycles-frontend   #    0.86% frontend cycles idle     (37.48%)
   101,216,673,486      stalled-cycles-backend    #   68.14% backend cycles idle      (37.54%)
   332,712,119,256      instructions              #    2.24  insn per cycle         
                                                  #    0.30  stalled cycles per insn  (37.59%)
    21,979,315,206      branches                  #  612.838 M/sec                    (37.55%)
       226,960,797      branch-misses             #    1.03% of all branches          (37.51%)
   116,473,309,464      L1-dcache-loads           # 3247.566 M/sec                    (37.47%)
     8,746,305,456      L1-dcache-load-misses     #    7.51% of all L1-dcache hits    (37.44%)
   <not supported>      LLC-loads                                                   
   <not supported>      LLC-load-misses                                             

       8.098330846 seconds time elapsed

      34.293568000 seconds user
       1.560401000 seconds sys
```

**10K dataset**
`time ./a.out measurements_10k.txt > log_feed.txt`
real	0m53.773s
user	3m41.558s
sys	0m3.801s

`perf stat -d ./a.out measurements_10k.txt`
```
 Performance counter stats for './a.out measurements_10k.txt':

        223,792.88 msec task-clock                #    4.141 CPUs utilized          
            20,923      context-switches          #    0.093 K/sec                  
                 6      cpu-migrations            #    0.000 K/sec                  
           954,607      page-faults               #    0.004 M/sec                  
   927,153,618,406      cycles                    #    4.143 GHz                      (37.49%)
    36,032,937,984      stalled-cycles-frontend   #    3.89% frontend cycles idle     (37.50%)
   378,911,873,059      stalled-cycles-backend    #   40.87% backend cycles idle      (37.51%)
 1,085,040,015,599      instructions              #    1.17  insn per cycle         
                                                  #    0.35  stalled cycles per insn  (37.51%)
   185,978,223,527      branches                  #  831.028 M/sec                    (37.50%)
     6,741,542,863      branch-misses             #    3.62% of all branches          (37.50%)
   475,547,270,304      L1-dcache-loads           # 2124.944 M/sec                    (37.50%)
    36,951,109,992      L1-dcache-load-misses     #    7.77% of all L1-dcache hits    (37.49%)
   <not supported>      LLC-loads                                                   
   <not supported>      LLC-load-misses                                             

      54.044563006 seconds time elapsed

     220.074017000 seconds user
       3.692581000 seconds sys

```

**Reply by lehuyduc** [reply]
I don't have simple access to Intel PC, so I can't test on Intel, unfortunately :(

**Reply by dzaima** [reply]
Thanks! That does just look like it's taking ~4.4x more instructions & branches compared to what it does on my PC for the regular dataset, and ~35x more on the 10k dataset. Which is actually useful datapoint - the fallback code in my solution is a boring `std::unordered_map<std::string, int64_t*>`, which the 10k dataset mostly spends time in, perhaps your systems have a significantly less efficient stdlib. I just really wanted to avoid having to write a proper hashmap.. I'll see about writing a custom one tomorrow, that's a good idea anyways

**Reply by lehuyduc** [reply]
Btw, your code is wrong when 4 threads are used instead of 8. Maybe a simple offset bug somewhere:
```
time THREADS_1BRC=4 ./a.out measurements.txt > log_feed.txt

real	0m33.074s
user	0m47.501s
sys	0m0.911s

sha256sum log_feed.txt 
89162cdb78505a14891cacf920702b086ecb4b9b04aeb08179d733efc96a1159  log_feed.txt
```

Running with 8 threads output the expected sha256sum: a55d0d9d02661c33538f2e11bb86f1825a5f015d6dd3645416ec71bc50099ee5

**Reply by dzaima** [reply]
oh geez, still had some `int`s left where 64-bit ints should be.. fixed that. That's what I get for not bothering to test on the full dataset much..

**Reply by dzaima** [reply]
quadruple-geez - so that bug meant that my solution on the full input has been giving the low input bound for each thread at ≤2B bytes while having the expected upper bound, thus running over parts of the input multiple times, which explains things.. (explains everything perhaps? the 10k dataset is even larger, so with my solution being, uh, O(n^2) (in addition to being very wrong) its even more significant slowdown makes sense. so its been me being stupid all along :/)

**Reply by lehuyduc** [reply]
@dzaima Glad to help you find some bugs! Can you generate some random data and test my code too? You can try running it with any different number of threads like `./run_cpp.sh 17` It's hard to test your own code :D

**Reply by dzaima** [reply]
@lehuyduc your solution on some of my tests has been giving results off by 0.1 due to rounding differences, making it hard to compare (some of the tests I've made do just have few samples per city leading to exact half-point values). With [this](https://dzaima.github.io/paste/#0pVPbbqMwEH3nK0Z5iKBgLgnppiWR@rCfsG/dChkwrCVjIjCI7qr/vr4QkjbJrqoiJHtmzpyZObYLWpaAUEUF4CDK2jwtKccsHTCjhZ8fDpBddVuUF2SEzUOWxWXm@@vtipT3G4jC8D6OLYTQDT7Ldd1bnE9PgNbx2tuCq5YoBOkZGlrAAbeYMcJSXFUtqbAgKRtWNuUCBC08UBt@Cl44FNqxAODNkr@FStZgAW3T8@JHE30nOa0xs42X93VGWgf@WEgmQEtE33LoRPH4qBNsA4A7OawfOhDoNbHcoukzRi5YJ/eoGF3QnVHYG0JmGEcn0aES7BHJ4B5CfwPLJdCdLEBdV4enTmhg6qlBpDoMCyrpFW3R1/VromWMt9@8KAI33j540UrpqAA1plyLhtsq9yD/hds7tR@eX7Q66ssb3klAL5qlrNj1TMhmzaZ7pi/JNVwnZBMNTzmuyYz2S9p24ia@OwE7IoNFYvQGMMeAh0ohPog5qeloBr/r60l@uRhXzoXzgal@rfF4hcskqJjhuJJH@e08GZvzXJOnpvtE4@cdnzP8t@FL/L8blfjjMZQN7HbvD0zai/1CLYZI2YG21SgnS7c1Vw6CL3Od2bMs831Rj4HC7nj1/I7@JrYDCCJnKrzwYJFMj/o42OLtJ5fOvw#y0nMS9dQT8lMS1PXUU8uKFDXtAYA) patch to make its rounding match mine (which I'm now noticing sometimes gives `-0.0`.. ¯\\\_(ツ)\_/¯), and also just switching to compare sum instead of average, it does pass the tests I've written!
(minor edit for fun: with the 1.38GB test, `./run_cpp.sh 2` is ~0.64s, `./run_cpp.sh 256` - 1.1s; whereas my solution "scales" "less worse" with `THREADS_1BRC=2 ./a.out` being ~0.58s and `THREADS_1BRC=256 ./a.out` - 0.77s; very pointless to run 256 threads (or forks as it may be in the case of my solution) on a 2c4t machine of course)

**Reply by lehuyduc** [reply]
@dzaima oh, for that rounding problem, can you check the actual sum/count values? To see if we're rounding correctly. Thanks for testing!

**Reply by dzaima** [reply]
I did check the sums, and they did match mine. Didn't bother with counts, that's much more difficult to screw up if the sums work.

**Reply by lehuyduc** [reply]
I notice that you use `fork()` instead of threads. I wonder how that affects `munmap` time. Will it be evenly divided per child process instead of being linear?

**Edit:** oh no it seems `fork()` really allows parallel munmap unlike thread. This will cause an arm race 😭

**Reply by dzaima** [reply]
I did mention fork-parallel unmapping [here](https://github.com/gunnarmorling/1brc/discussions/138#discussioncomment-8182046); I `mmap` 32MB chunks too, so I get to start doing some effective-`munmap` after just going through 32MB, interleaved through the whole execution, separately for each fork. You can change the `constexpr bool global_mmap = false;` to `true` to see how timings are impacted by having just one massive `mmap` before forking instead.

**Reply by lehuyduc** [reply]
[result_dzaima.txt](https://github.com/gunnarmorling/1brc/files/13998559/result_dzaima.txt)
[result_dzaima_10k.txt](https://github.com/gunnarmorling/1brc/files/13998560/result_dzaima_10k.txt)

I've just benchmarked your code again, with both default + 10K dataset. All thread range from 1 => 32. Commit 
fa6db52

**Good news:** it's super fast now on the default dataset!! It's still slow on the 10K dataset, but I assume it's because you haven't worked on it yet. Also `global_mmap = true` doesn't decrease any performance.

**Bad news:** By using `fork()` instead of threads, your program is getting the exact same speedup as creating a child process, do everything there, then `kill` the parent process. I tested that by implementing that trick in your code, but there's no speedup. That means, if you want a fair comparison with everyone else, we have to subtract everyone's `munmap` + free memory time (because if `fork()` is legal then everyone can just implement this trick and get the same amount of speedup). Actually I'm not sure about this. Like, there's no official rules about this.

```
void handleSignal(int signum) {
    if (signum == SIGUSR1) {
        printf("Received SIGUSR1 signal. Exiting.\n");
        exit(0);
    }
}

int main(int argc, char* argv[]) {
  signal(SIGUSR1, handleSignal);

  int pid = fork();
  int status;
  if (pid == 0) {
    return do_everything(argc, argv);
  } else {
    waitpid(pid, &status, 0);
  }
}
```

Amd 2950x
**Default dataset**
```
1 thread
real    0m8.806s
user    0m8.299s
sys     0m0.500s

16 thread
real    0m0.663s
user    0m9.501s
sys     0m0.608s

32 thread
real    0m0.549s
user    0m15.816s
sys     0m1.019s
```

**10K dataset**
```
1 thread
real    0m53.456s
user    0m52.469s
sys     0m0.956s

2 thread
real    0m27.355s
user    0m53.389s
sys     0m0.952s

8 thread
real    0m7.107s
user    0m54.899s
sys     0m1.076s

real    0m7.092s
user    0m54.435s
sys     0m1.059s

16 thread <= performs better than 32 threads
real    0m3.927s
user    0m59.005s
sys     0m1.204s

32 thread
real    0m5.938s
user    3m0.619s
sys     0m2.054s```

**Reply by dzaima** [reply]
Right, I haven't worked at all on the 10k dataset; I'd probably want to make a separate thing for it if I wanted to do it (a trivial thing would be making the hashmap have more than 4096 slots, but then it might slot in a different cache level).
(**edit**: actually, currently the hashmap size is a dynamic value within the SIMD code, so I could theoretically adapt the hashmap size to the input)

I don't really understand all the `munmap` business. Is it that the kernel cleans up the fork in the background, potentially after the main program has finished? If so, I don't think that should apply to my solution with `global_mmap = false`, as each fork, upon completing, has only 32MB mapped; the vast majority of the input is already removed from the address space.
